### PR TITLE
WIP: Constraints and Loss Class

### PR DESCRIPTION
**Pull request by mayou36**
Added a LossClass (Interface, Base and an implementation, UnbinnedNLL)
Added constraints support

+ a lot of small changes

TODOs:
- [x] `__add__` for loss
- [x] move unbinned_nll function body inside nll
- [x] pdf_base: auto_convert to fixed parameter if not parameter

### CODE REVIEW

File: tests/test_loss.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+import zfit.settings
+from zfit.core.loss import unbinned_nll, UnbinnedNLL
+
+mu_true = 1.2
+sigma_true = 4.1
+
+test_values_np = np.random.normal(loc=mu_true, scale=sigma_true, size=1000)
+
+low, high = -24.3, 28.6
+mu1 = FitParameter("mu1", ztf.to_float(mu_true)-0.2, mu_true - 1., mu_true + 1.)
```
Comment (apuignav) [inline tests/test_loss.py]: Why do we need `ztf.to_float` here?

File: zfit/core/loss.py [hunk 2]
```diff
@@ -1,25 +1,220 @@
...
+
+    def errordef(self, func):
+        raise NotImplementedError("For this simple loss function, no error calculation is possible.")
+
+    def _loss_func(self, pdf, data, fit_range, constraints=None):
+        loss = self._simple_func
+        return loss()
+
+
+def make_loss(func):
```
[the remark this answers is not in the document]
Comment (mayou36) [reply zfit/core/loss.py]: It's a short-cut I've added, let's see if needed. But while the loss object is very useful for error estimation, it is actually not needed to just do a minimization. E.g. if you just build a TF function and want it to be minimized, `make_loss` simply converts that to a trivial loss function that can be minimized. And raises errors if it tries to go beyond the borders like calculating errors.

I am not sure whether this should be there (or be high level) but it's at least very useful for testing (though then it could be moved to the tests) and probably to do a simple minimization. Let's say, it's a wrapper around a simple TF Tensor that can be used for minimization.

More of a side-product. Don't like it?

File: zfit/core/loss.py [hunk 3]
```diff
@@ -1,25 +1,220 @@
...
+    return sigma
+
+
+class UnbinnedNLL(BaseLoss):
+    _name = "UnbinnedNLL"
+
+    def _loss_func(self, pdf, data, fit_range, constraints):
+        return unbinned_nll(pdf=pdf, data=data, fit_range=fit_range, constraints=constraints)
+
+    def errordef(self, sigma):
```
[the remark this answers is not in the document]
Comment (mayou36) [reply zfit/core/loss.py]: well, sigma. If sigma is one, it's one. But in general it's sigma, right? (except of a nasty factor of 2 if you have -2sum(log(x))

File: zfit/core/loss.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (apuignav) [reply zfit/core/loss.py]: I was wondering why not simply do `minimize(SimpleLoss(x))`?

File: zfit/core/loss.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (apuignav) [reply zfit/core/loss.py]: Yeah, I'm not sure I would want to leave this freedom. Errordef is used by minuit in special ways, I think, so I think this should always be 1 sigma.

File: zfit/core/loss.py [hunk 4]
```diff
@@ -1,25 +1,220 @@
...
+
+def errordef_nll(sigma):
+    return sigma
+
+
+class UnbinnedNLL(BaseLoss):
+    _name = "UnbinnedNLL"
+
+    def _loss_func(self, pdf, data, fit_range, constraints):
+        return unbinned_nll(pdf=pdf, data=data, fit_range=fit_range, constraints=constraints)
```
[the remark this answers is not in the document]
Comment (mayou36) [reply zfit/core/loss.py]: True, there is no direct need... Reminds me that I forgot to add the `__add__`, will be done.

File: zfit/core/loss.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (mayou36) [reply zfit/core/loss.py]: My idea was to probably use it inside a pdf as an argument converter if it's a Tensor. But I don't like the idea too much. So it's just a little factory. But no real need. Remove?

File: zfit/core/loss.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (apuignav) [reply zfit/core/loss.py]: Now that we are on it... What are we planning to do with `_name`?

File: zfit/core/loss.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (apuignav) [reply zfit/core/loss.py]: I would say remove for the time being.

File: zfit/core/loss.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (mayou36) [reply zfit/core/loss.py]: Sure? At least in the iminuit docs, they describe it as "sigma: number of σ error."

Shouldn't `errordef` basically be the converter from n sigma to loss function?
e.g. if you use the 2nll loss function, then 1 sigma -> 0.5?

But I may miss something, should read up on it

File: tests/test_loss.py (hunk 1 above)
Comment (mayou36) [reply tests/test_loss.py]: Currently, Gauss actually expects a parameter and does not convert the value therefore. But this should be changed! You're right

Comment (mayou36) [reply tests/test_loss.py]: Or better, autoconverted to fixed parameters, right?

Comment (mayou36) [reply tests/test_loss.py]: That's a little bit the point here: it does not expect fixed parameters *per se* (it's a pdf in the end). But to use it as a pdf to constrain a value, we need fixed parameters.

So to be precise: whenever anything expects a *fixed/floating* parameter and get's a float, convert to a fixed param.

I agree on it!

Comment (apuignav) [reply tests/test_loss.py]: Ah yeah, I was sloppy: anything that expects a parameter and gets a float, is converted to a fixed parameter (unnamed?).

Comment (mayou36) [reply tests/test_loss.py]: unnamed -> somehow auto-named? Like "AUTOPARAM_fixed" or something like this?

**Comment by mayou36** [comment]
Except of the few discussions that are open, this is ready from my side.

One thing I did differently: I did not fully put the `unbinned_nll` function inside the class body but kept is as a function outside (though split it in two and renamed with leading underscore).

This can be useful to add more constraints or build an extended likelihood etc.

Anyway: in the (public) namespace, there is only UnbinnedNLL available.

As I produced a few merge conflicts, let me know when it's ready and I merge it locally, ok? To get rid of the conflicts

**Comment by apuignav** [comment]
> What's the expected behavior of `__add__`?
> 
> 1. UnbinnedNLL + UnbinnedNLL = a) UnbinnedNLL? b) SimultaneousLoss?
> 2. UnbinnedNLL + OtherLoss = a) Error? (because why?) b) SimultaneousLoss?
> 3. Anyway, create SimultaneousLoss to allow mixing? (e.g. if a) in 2.)
> 
> My take: 1. a), 2. a), 3. yes (not yet done)

I think 1. a), 2.a), 3 not sure (I don't think it makes much sense). In a more general way, any `__add__` should add losses of the same type I think.

File: tests/test_loss.py (hunk 1 above)
Comment (apuignav) [reply tests/test_loss.py]: Do we need names? If so, autonamed then!
